Device and Platforms tests

cojoj opened the pull request:
@czechboy0 I hope you're back from holidays, full of energy and ready to work 😆

#### New batch with *Device* tests.

First of all, I've added missing fields/properties to Device model - please check them and let me know what do you think and if things marked as `Enum?` in comments should be moved to _DeviceSpecyfication_ file.

One thing that concerns me about _cassetes_ is that we're only testing *gold path* and skipping any possible error paths. Do you think we should create test cases and test failing paths as well? If so, how would you write those tests? Record another casette, write mocks or stubs? As we leave it as it is now we won't get full coverage of the code...

cojoj commented:
#### Platforms tests

buildasaur commented:
Result of integration 1
Integration took 8 seconds.
? build errors: build-failed

buildasaur commented:
Result of integration 2
Integration took 44 seconds.
Perfect build! All 25 tests passed. :+1:

buildasaur commented:
Result of integration 1
Integration took 6 seconds.
? build errors: build-failed

buildasaur commented:
Result of integration 1
Integration took 47 seconds.
Perfect build! All 25 tests passed. :+1:

czechboy0 commented on XcodeServerSDK/Server Entities/Device.swift:
```diff
@@ -10,21 +10,55 @@ import Foundation
+        self.modelName = json.stringForKey("modelName")
+        self.deviceECID = json.optionalStringForKey("deviceECID")
+        self.modelUTI = json.stringForKey("modelUTI")
+        if let proxyDevice: NSDictionary = json.optionalForKey("activeProxiedDevice") {
```
You can just use `if let proxyDevice = json.optionalDictionaryForKey("activeProxiedDevice") {`

cojoj replied:
My mistake. Earlier I was trying to achieve this other wa around and this is a small _artifact_...

czechboy0 commented on XcodeServerSDK/Server Entities/Device.swift:
```diff
@@ -10,21 +10,55 @@ import Foundation
+    public let deviceECID: String?
+    public let modelUTI: String
+    public let activeProxiedDevice: Device?
+    public let docType: String
```
`docType` is an internal Xcode Server key used in CouchDB to identify the... surprise surprise... document type. If you want to add it, it should be added to `XcodeServerEntity` instead, because it's also present in entities like bots and integrations.

cojoj replied:
Yeah, I've noticed that later and I'll move it to `XcodeServerEntity`.

czechboy0 replied:
:+1:

czechboy0 commented on XcodeServerSDKTests/Casettes/get_devices.json:
```diff
@@ -0,0 +1,366 @@
+              "isServer" : false,
+              "retina" : false,
+              "tinyID" : "76AFAC6"
+            }
```
Too bad we don't have any real iOS devices in the list here :( We might even remove most and just keep one per type to keep these JSON canned responses maintainable.

cojoj replied:
I think it's not a good sign to share with public identifiers and so on... 😉 I guess we've no other possibility to achieve this test with simulators. BTW, I think that any real device should be removed from this casette. We can manually prepare it with fake ID's.

czechboy0 replied:
Good point, feel free to generate a random UUID for the real devices.

czechboy0 commented:
Let me know when you push the changes so that we can merge this :wink:

cojoj commented:
Ok, changed something, added other things and here's the result... 😉

You haven't answered how to handle *failing paths* in test cases. Got any idea?

buildasaur commented:
Result of integration 2
Integration took 34 seconds.
Perfect build! All 25 tests passed. :+1:

czechboy0 commented:
Yeah the failing paths are interesting. There is not that much code different for each endpoint. Most of the logic is shared. So I'd rather write tests for `XcodeServer`'s `sendRequestWithMethod` instead of writing very similar failing paths for each API call. Whadaya think?

czechboy0 commented:
Cool, thanks as always, @cojoj!

cojoj commented:
Exactly, one **complete** test case for `sendRequestWithMethod()` can solve all of our problems with specyfic test cases but to write this test we need to create stub so we can imitate the failing response. You know what I mean?

czechboy0 commented:
Yup, feel free to use any of the available ones, but it needs to be clear we're testing `sendRequestWithMethod`, not e.g. `getBots`, if you know what I mean.

cojoj commented:
I guess I'd be hard to achieve as `sendRequestWithMethod()` has `internal` accessory flag...

czechboy0 commented:
Should be fine with `@testable import XcodeServerSDK` in the test file, shouldn't it?

cojoj commented:
Probably... I'll try to write _something_

czechboy0 commented:
C'mon, http://natashatherobot.com/swift-2-xcode-7-unit-testing-access/

cojoj commented:
Hah, I think that testing `sendRequestWithMethod()` won't solve everything. I think what needs to be tested is `createRequest()` and `sendRequest()` (which leads to `dataTaskWithRequest()`).

czechboy0 commented:
Why? Just call it on an `XcodeServer` you get from `getRecordingXcodeServer`, just like with the other tests.

The only thing `sendRequestWithMethod` does is basically call `createRequest` and right away `sendRequest` on it :blush: Or am I missing something?

cojoj commented:
Right, it doesn't have wide range of responsibilities but you have to fake HTTP responses to check the failable path eg. To test this code (from `sendRequestWithMethod()`):
```swift
uard let data = data else {
    //no body, but a valid response
    completion(response: httpResponse, body: nil, error: nil)
    return
}
```
You have to return `nil` body from `self.session.dataTaskWithRequest()` and to do this you need to provide prepared `NSURLRequest` which will return proper response...

Correct me I you think that I'm spinning a yarn...

czechboy0 commented:
Ok this makes sense. I guess we should test with separate `createRequest` and `sendRequest` in this case and with `sendRequestWithMethod` for different cases, such as with different error returned, server unavailable, wrong returned content types etc.

cojoj commented:
Cool!

Anyway, I hate the idea that all in all we'll have to create manually some fake data 😓

czechboy0 commented:
:crying_cat_face:
